Add standalone bevy_entity_inspector and bevy_transform widget for drawing on the UI

jbuehler23 opened the pull request:
# Add Entity Inspector with Component Grouping and Event-Driven Updates

## Objective

Bevy currently lacks a comprehensive, user-friendly entity inspector for runtime debugging and development. Developers need a way to:

- **Inspect entities and components** in real-time during development
- **Understand component organization** by seeing which crates contribute which components
- **Connect to remote applications** for debugging without modifying the target application
- **Navigate large entity hierarchies** efficiently with expandable tree views
- **See only relevant data** with clear visual indicators of what can be expanded

The `bevy_entity_inspector` crate provides a first-party entity inspector with modern UX patterns, designed specifically for Bevy's component model and reflection system.

## Solution

### Component Grouping by Crate

Components are automatically organized by their crate origin for better comprehension:

```
Entity (42)
├── bevy_transform
│   ├── Transform (expandable - has field data)
│   │   ├── translation: Vec3(0.0, 0.0, 0.0) 
│   │   ├── rotation: Quat(0.0, 0.0, 0.0, 1.0)
│   │   └── scale: Vec3(1.0, 1.0, 1.0)
│   └── GlobalTransform
├── bevy_render
│   ├── Visibility (dimmed - no expandable data)
│   └── ViewVisibility (dimmed - no expandable data)
└── my_game
    └── Player
        ├── health: f32(100.0)
        └── level: u32(1)
```

### Property Panel Features

When you select a component (e.g., `Transform`), the property panel immediately displays:

- **Component header**: Clear component name and type
- **All reflected fields**: Instantly visible without drilling down
- **Color-coded display**: Field names in light blue, values in white
- **Structured layout**: Each field in its own container for readability


### Visual Design System

- **Color-coded hierarchy**: Different colors for entities, crate groups, components, and fields
- **Expand Indication**: Reduced opacity for components without field data
- **Disclosure triangles**: Clear expand/collapse indicators  
- **Responsive design**: Scrollable interface with hover effects (though mouse wheel doesn't work...)

### Remote Inspection Support

- **`bevy_remote` integration**: Connect to any Bevy app with remote plugin enabled
- **Reflection-based**: Works with any component that implements `Reflect` and `ReflectDeserialize`
- **Graceful degradation**: Components without reflection support still appear in tree

## Testing

### Basic Inspector Example
```bash
# Run basic inspector (empty until data source configured)
cargo run --example inspector -p bevy_entity_inspector
```

### Remote Inspection Testing  
```bash
# Terminal 1: Start target application with remote plugin
cargo run --example cube_server -p bevy_entity_inspector --features remote

# Terminal 2: Start inspector connected to remote app  
cargo run --example inspector -p bevy_entity_inspector --features remote
```

### More Verifcation and Testing

**Component Coverage Verified:**
-  **Transform hierarchy**: `Transform`, `GlobalTransform`, `TransformTreeChanged`
-  **Rendering components**: `Mesh3d`, `MeshMaterial3d`, `RenderEntity`, `SyncToRenderWorld`  
-  **Lighting components**: `DirectionalLight`, `PointLight`, `SpotLight`, `ShadowSettings`
-  **Camera components**: `Camera`, `Camera3d`, `CameraRenderGraph`, `Projection`
-  **Visibility components**: `Visibility`, `ViewVisibility`, `InheritedVisibility`
-  **Custom components**: User-defined components with reflection support

**Performance Testing:**
-  **Large entity counts**: Tested with 1000+ entities in remote mode
-  **Component field expansion**: Deep component field hierarchies
-  **Real-time updates**: Live component value changes during gameplay


**UI/UX Validation:**
-  **Expansion indicators**: Clear visual feedback for expandable vs non-expandable components
-  **Crate grouping**: Proper organization of Bevy engine vs user components  
-  **Tree navigation**: Smooth expand/collapse with state preservation
-  **Performance**: Responsive UI even with large component trees


## Showcase
With reflectable fields:
![image](https://github.com/user-attachments/assets/456d1789-1ceb-4718-b5a2-10d6da63b765)

Just entity data:
![image](https://github.com/user-attachments/assets/af6fdadb-04ca-4316-9589-6625f3222be2)

No reflectable fields on the selected component:
![image](https://github.com/user-attachments/assets/4b2915ec-89d5-4afa-8511-159beda9d23e)

jbuehler23 committed:
Add Component Viewer pane and integrate bevy_remote functionality by pressing "space" key

jbuehler23 commented on bevy_editor_panes/bevy_component_viewer/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+    fn build(&self, app: &mut App) {
+        app.register_pane("Components", setup_pane).add_systems(
+            Update,
+            connect_to_remote.run_if(input_just_pressed(KeyCode::Space)),
```
i added this in just in case it was a timing issue with the http server not fully being up yet by the time the first Update cycle kicks in. Still no luck with this though!

jbuehler23 committed:
Refactor connect_to_remote function to use IoTaskPool for asynchronous requests and improve logging

jbuehler23 committed:
Add JSON deserialization for the bevy_remote responses, as well as implement a polling mechanism. Also will update the "Component Viewer" UI with the latest entity data based on the currently selected entity as well!

jbuehler23 committed:
Code cleanup

jbuehler23 committed:
Refactor changes to create a new bevy_transform widget, and bevy_entity_inspector (using original bevy_component_viewer functionality). idea is to make this modular, so that it can be used for any bevy app

jbuehler23 committed:
Major refactor to use rows of entity data, and reflected components to deserialize in the UI

jbuehler23 committed:
feat: Refactor Bevy Entity Inspector with Event-Driven Architecture and Component Grouping

- Introduced an event-driven architecture replacing hash-based change detection with granular `InspectorEvent` system.
- Organized components by crate name in the tree UI for better clarity.
- Implemented a visual styling system with distinct colors for entities, crate groups, components, and fields.
- Enhanced UI organization to resemble professional ECS inspectors like Flecs.
- Improved change tracking efficiency by separately tracking added, removed, and updated entities.
- Updated `EntityInspectorRow` to include a `data_hash` field for change detection.
- Simplified component names to "crate::Type" format for remote inspection.
- Added comprehensive documentation for all public APIs and performance notes.
- Fixed runtime panic issues and improved memory efficiency.
- Created a theme configuration for the inspector with dark and light themes.
- Developed disclosure triangle UI components for collapsible tree nodes.
- Implemented a tree view UI for displaying hierarchical data with selection and expansion features.
- Added inspector panel and field widgets with customizable properties.
- Established integration tests for the inspector plugin functionality.

jbuehler23 committed:
chore: Remove outdated CHANGELOG and PR_DESCRIPTION files after refactoring the Bevy entity inspector

jbuehler23 committed:
docs: add more docs to fix warnings

jbuehler23 committed:
feat: Add entity inspector with property panel and event handling

- Implemented a property panel example for displaying entity components.
- Created event system for entity inspector updates, including entity and component events.
- Developed reflection utilities for extracting and formatting component data.
- Built tree structures for the inspector UI, grouping components by crate.
- Designed a property panel UI for displaying detailed component information.
- Established UI management systems for handling inspector events and tree updates.
- Added camera setup for inspector rendering.
- Implemented tree node selection handling to update the property panel.

jbuehler23 committed:
refactor: Update UI layout and improve property panel width for better usability

jbuehler23 committed:
Merge branch 'main' of https://github.com/bevyengine/bevy_editor_prototypes into investigate-brp-usage

jbuehler23 committed:
feat: Add missing dependency for bevy_remote in Cargo.toml

alice-i-cecile commented on bevy_widgets/bevy_transform/src/lib.rs:
```diff
@@ -0,0 +1,67 @@
+//! A Bevy plugin for displaying and editing transform data in a UI widget.
+//! This plugin provides a UI template that displays the translation, rotation, and scale of a transform in a structured format.
+//! It uses the `bevy_i_cant_believe_its_not_bsn` crate to create the UI template.
```
Remember to add doc links: they're helpful for navigating, but also extremely useful because they break when the references go stale, reminding people to update the docs.

jbuehler23 replied:
i've actually removed this, but i'll check through the other docs to add links

alice-i-cecile commented on bevy_widgets/bevy_transform/src/lib.rs:
```diff
@@ -0,0 +1,67 @@
+//! A Bevy plugin for displaying and editing transform data in a UI widget.
```
I feel like we can split out this entire crate / piece of work, making this PR easier to review and getting it in faster.

jbuehler23 replied:
Totally agree - this is actually leftover from when I initially started working on _only_ displaying Transform values until BRP changes went out

alice-i-cecile commented on crates/bevy_entity_inspector/src/events.rs:
```diff
@@ -0,0 +1,302 @@
+/// These events are typically emitted by data source plugins (like the remote inspection plugin)
+/// and consumed by the main event handler to update the tree UI.
+#[derive(Event, BufferedEvent, Debug, Clone)]
+pub enum InspectorEvent {
```
A comment here about why this is an enum, rather than multiple distinct events would be useful. I think it will probably make serde a bit nicer, and it's more discoverable, and they all fit together because they force a rebuild but I was initially quite confused.

jbuehler23 replied:
Yes, totally agree - I've gone back and added much more in-depth documentation.

alice-i-cecile commented on crates/bevy_entity_inspector/src/ui/tree.rs:
```diff
@@ -0,0 +1,600 @@
+/// A node in the tree structure
+#[derive(Component, Clone, Debug)]
+pub struct TreeNode {
+    /// Unique identifier for the node
```
This could use more color: how are these values populated? How does this differ from `label`?

jbuehler23 replied:
fully agree - i've added some more in-depth docs just now!

alice-i-cecile commented on crates/bevy_entity_inspector/src/ui/tree.rs:
```diff
@@ -0,0 +1,600 @@
+        ))
+        .id();
+
+    // Add disclosure triangle or spacer for visual indication
```
```suggestion
    // Add disclosure triangle for visual indication
```
I think this is clearer?

alice-i-cecile reviewed (commented):
Very nice! I still need to try this out hands-on. I would prefer if the two tools were split into distinct PRs, but I won't block on it.

I've left a few suggestions, but nothing blocking at this stage. Overall quality is good: this is something I'm comfortable refining.

jbuehler23 committed:
Update tree.rs

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

jbuehler23 commented:
> Very nice! I still need to try this out hands-on. I would prefer if the two tools were split into distinct PRs, but I won't block on it.
> 
> 
> 
> I've left a few suggestions, but nothing blocking at this stage. Overall quality is good: this is something I'm comfortable refining.

Nice! Thank you for the review. There's definitely some UX improvements we need to refine, and also there is some flickering happening every time we get the latest updates from BRP. I'll collate a list of refinements we should make as well as separate issues

jbuehler23 committed:
refactor: Remove bevy_transform plugin and associated files

jbuehler23 committed:
Merge branch 'investigate-brp-usage' of https://github.com/bevyengine/bevy_editor_prototypes into investigate-brp-usage

jbuehler23 committed:
docs: Enhance documentation across entity inspector modules with related links and detailed descriptions

jbuehler23 committed:
docs: Improve documentation for tree structure components and their interactions

jbuehler23 committed:
Merge branch 'main' into investigate-brp-usage

jbuehler23 committed:
refactor: Reorder imports in cube_server.rs for clarity
style: Clean up whitespace in remote.rs for consistency

jbuehler23 committed:
Merge branch 'investigate-brp-usage' of https://github.com/bevyengine/bevy_editor_prototypes into investigate-brp-usage

jbuehler23 committed:
fix clippy!

jbuehler23 committed:
refactor: Change ui module from public to private

jbuehler23 committed:
docs: Fix formatting in inspector example documentation

jbuehler23 committed:
fix: Correct documentation formatting and improve clarity in reflection and UI systems

syborg64 committed:
Update entity_inspector to workspace version of bevy (#266)

* Add advanced entity outline calculations (#222)

* Add advanced entity outline calculations

The changes add sophisticated AABB (Axis-Aligned Bounding Box)
calculations for accurately drawing outlines around selected entities,
handling various entity types like meshes and sprites, and properly
accounting for hierarchies and transformations.

* minor changes

* Extract outline gizmo queries into SystemParam struct

The change extracts multiple queries used in the outline gizmo system
into a dedicated `OutlineGizmoQueries` struct that implements
`SystemParam`

* fmt

* Rename outline gizmo plugin to selection box plugin

* Rename outline gizmo to selection box

The changes are focused on renaming the outline gizmo functionality to
selection box, which includes renaming the module, component, functions
and UI text for better clarity and consistency.

* Launcher code quality improvements (#223)

* Simplify `selection_box::transform_aabb` (#224)

* Fix panel layout exploding when panels overflow (#227)

* Fix 3d viewport picking passthrough and camera input (#226)

* Fix 3d viewport picking passthrough and camera input

* Adjust docs

* Update to rust edition 2024 (#228)

* `cargo fix --edition` minus some unwanted changes

* Change edition to 2024

* `cargo format --all`

* Allow loading gltf/glb files from outside the assets directory (#231)

* Improve selection (#230)

* Improve selection

* Update crates/bevy_editor_core/src/selection.rs

Co-authored-by: JMS55 <47158642+JMS55@users.noreply.github.com>

---------

Co-authored-by: JMS55 <47158642+JMS55@users.noreply.github.com>

* Update .gitignore (#229)

* Depend on cart's BSN branch (#232)

* Update to `0294fa89db4e213ac8e186eed6966a7ea7459d1a`

This is the commit right before cart's BSN PR

* Depend on cart's BSN branch

* argh

* this time for real

* Update to latest commit + enable bevy_feathers + remove unintended changes to Cargo.toml

I'm keeping the `resolver = "3"` >:D

* Humble Beginnings for the `ActionRegistry` (#234)

* Port Viewport Panes to BSN and feathers (#233)

* Port Properties Pane to BSN and Feathers (#235)

* Port Properties Pane to BSN and Feathers

* Adjust editor core prelude

* Use Theme Colors for Viewgizmo (#237)

* Use theme colors for viewgizmo

* clipperry

* Vendor `bevy_transform_gizmo` (#236)

* Vendor `bevy_transform_gizmo`

Co-authored-by: Aevyrie <aevyrie@gmail.com>

* Remove stub

* Rename

* It compiles

* Kinda working + cleanup

* cleanup + fix parenting example

* bundle to required components

* Switch to observers + Finally working properly now :)

* refactoring + cleanup + docs

* more refactoring + finish writing docs

* refactor

* Delete `.gitignore`

* fixes

* Cleanup unwraps and errors + Add proper docs to crate root + auto add `MeshPickingPlugin` + add back `With<GizmoTransformable>`

* typooo

---------

Co-authored-by: Aevyrie <aevyrie@gmail.com>

* Document the Lucide placeholder icon font (#241)

* docs: add `styles.md` to design book

calling it the "Themes and Styles" page after this:
https://m3.material.io/styles

please feel free to change it if you want :)

* docs(styles.md): add info about icon set (Lucide)

* Update Bevy dependency (#242)

* Add support for selecting multiple entities (#240)

* Add support for selecting multiple entities

* Switch to `UniqueEntityVec`

* Rework removal

* Add `FromIterator` impl

* Add tests

* Fix up tests

* Remove brittle id gen in tests

* Enable Transform Gizmo (#244)

* Rework action keybinding (#243)

* Rework keybindings

* Reorder keybinding constructor parameters

* Update Bevy (#245)

* Update Bevy

* Fix Rust 1.89 Lints

* Reduce the amount of debug info generated in CI

* Switch back to cart's branch

* Trigger CI

* Remove Manual Type Registrations (#246)

* Transform Gizmo/UX updates (#247)

* Implement comprehensive Bevy editor features including toolbar, scene tree, and properties pane enhancements

* Enhance Bevy toolbar and transform gizmos integration with improved UI elements, keyboard shortcuts, and snapping features

* remove readme

* Remove IMPLEMENTATION_PLAN.md from .gitignore

* refactor: add a simple shared color theme system for usage throughout the editor. this should be easy to bring in line with bevy_feathers

* fix: remove broken test

* refactor: update doc comment

* cargo format result

* Added basic sorting for asset browser (#256)

* Added basic sorting for asset browser

* Remove Ord implementation for Entry and move ordering to seperate functions

* Remove temporary fix I accidentally made permenant

* Prevent crash due to malformed heirarchy (#254)

* Removed unwrap from `update_render_target_size` to prevent crash on startup

* Removed unnecessary import

* Fix ci out of space (#263)

* Update ci.yaml

* Make sure no error throw

* Try a more aggressive fix

* Pass all checks

* default config looks fine

* swatinem not cache worksapce, need rebuild everytime

* Revert "swatinem not cache worksapce, need rebuild everytime"

This reverts commit e6552c11ef892f2b66fee5b8d48c875fd4ffffb7.

* Revert "default config looks fine"

This reverts commit 3cb28534948b8ba2683288afd007769285f51e11.

* Upgrade to 0.17 (#262)

* A minimal upgrade

* Correct the dependencies

* Continue

* Copy from @viridia (https://github.com/cart/bevy/commit/ab733198b8b59ec90194797f4a1d46a12043ebfc)

* Continue

* target() -> event().event_target()

* EventWriter/Reader -> MessageWriter/Reader & add_event -> add_message

* Fix display issue

* Pass all checks

* I hope this work

---------

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

* Moved projects.ron to comply with XDG base directory standards (#265)

* Moved projects.ron to comply with XDG base directory standards

* Changed unwrap_or to unwrap_or_else to avoid panic when XDG_DATA_HOME exists but HOME is not set

* Fixed issues with clippy

* fix: update to workspace version of bevy

* fix: doc links using rust paths instead of web links where possible

---------

Co-authored-by: okhai <57156589+okhaimie-dev@users.noreply.github.com>
Co-authored-by: Tim <JustTheCoolDude@gmail.com>
Co-authored-by: JMS55 <47158642+JMS55@users.noreply.github.com>
Co-authored-by: Aevyrie <aevyrie@gmail.com>
Co-authored-by: Barrett Ray <66580279+onkoe@users.noreply.github.com>
Co-authored-by: Joe Buehler <jbuehler23@gmail.com>
Co-authored-by: Freyja-moth <156322843+Freyja-moth@users.noreply.github.com>
Co-authored-by: Sieluna <seele.peng@gmail.com>
Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>
